webapp/src/accessibility.tsx

```diff
@@ -0,0 +1,118 @@
+/// <reference path="../../typings/globals/react/index.d.ts" />
+/// <reference path="../../typings/globals/react-dom/index.d.ts" />
+/// <reference path="../../built/pxtlib.d.ts" />
+
+import * as React from "react";
+import * as ReactDOM from "react-dom";
+import * as data from "./data";
+import * as sui from "./sui";
+
+type ISettingsProps = pxt.editor.ISettingsProps;
+
+export interface EditorAccessibilityMenuProps extends ISettingsProps {
+    highContrast: boolean;
+}
+
+export interface EditorAccessibilityMenuState {
+    highContrast?: boolean;
+}
+
+export class EditorAccessibilityMenu extends data.Component<EditorAccessibilityMenuProps, EditorAccessibilityMenuState> {
+
+    openJavaScript() {
+        pxt.tickEvent("accmenu.editor.openJS");
+        this.props.parent.openJavaScript();
+    }
+
+    selectLang() {
+        pxt.tickEvent("accmenu.editor.importdialog");
+        this.props.parent.selectLang();
+    }
+
+    toggleHighContrast() {
+        pxt.tickEvent("accmenu.editor.togglecontrast");
+        this.props.parent.toggleHighContrast();
+    }
+
+    goHome() {
+        pxt.tickEvent("accmenu.editor.home");
+        this.props.parent.exitAndSave();
+    }
+
+    componentWillReceiveProps(nextProps: EditorAccessibilityMenuProps) {
+        const newState: EditorAccessibilityMenuState = {};
+        if (nextProps.highContrast != undefined) {
+            newState.highContrast = nextProps.highContrast;
+        }
+        if (Object.keys(newState).length > 0) this.setState(newState)
+    }
+
+    shouldComponentUpdate(nextProps: EditorAccessibilityMenuProps, nextState: EditorAccessibilityMenuState, nextContext: any): boolean {
+        return this.state.highContrast != nextState.highContrast;
```

Comment:
isn't React supposed to do this already?

Reply:
Not the way we're using it. Since we have basically one AppState. Every time we call setState it calls a render on all components and their child components.
The way react is meant to be used is to have separate components each maintaining their own state. 
React still gives you the opportunity to override it's shouldComponentUpdate and have your own logic (as above) to when it should or shouldn't update the state, thus limiting the updates in our case.